lib/pos_mempool.go

```diff
@@ -0,0 +1,446 @@
+	return tb.txnsSet.Empty()
+}
+
+// GetIterator returns an iterator over the MempoolTx inside the TimeBucket.
+func (tb *TimeBucket) GetIterator() treeset.Iterator {
+	return tb.txnsSet.Iterator()
+}
+
+// GetTransactions returns a slice of MempoolTx inside the TimeBucket. The slice is ordered according to the
+// TimeBucketComparator.
+func (tb *TimeBucket) GetTransactions() []*MempoolTx {
+	txns := []*MempoolTx{}
+	it := tb.GetIterator()
+	for it.Next() {
+		if txn, ok := it.Value().(*MempoolTx); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+//============================================
+//	Fee-Time Bucket Math
+//============================================
+
+// computeFeeBucketBaseAndMultiplierFromGlobalParams takes the fee base rate and multiplier for the GlobalParamsEntry,
+// and returns them as big.Floats.
+func computeFeeBucketBaseAndMultiplierFromGlobalParams(globalParams *GlobalParamsEntry) (
+	_baseRate *big.Float, _bucketMultiplier *big.Float) {
+
+	feeBucketBaseRate := NewFloat().SetUint64(globalParams.FeeBucketBaseRate)
+	feeBucketMultiplier := NewFloat().SetUint64(10000 + globalParams.FeeBucketMultiplierBasisPoints)
+	feeBucketMultiplier.Quo(feeBucketMultiplier, NewFloat().SetUint64(10000))
+	return feeBucketBaseRate, feeBucketMultiplier
+}
+
+// computeTimeBucketMinFromIndex takes a fee range index and the GlobalParamsEntry, and returns the minimum fee rate
+// of this fee range.
+func computeTimeBucketMinFromIndex(index int, globalParams *GlobalParamsEntry) uint64 {
+	feeBucketBaseRate, feeBucketMultiplier := computeFeeBucketBaseAndMultiplierFromGlobalParams(globalParams)
+
+	// The first fee range has a fee rate of feeBucketBaseRate.
+	if index == 0 {
+		fee, _ := feeBucketBaseRate.Uint64()
+		return fee
+	}
+
+	// Compute feeBucketBaseRate * feeBucketMultiplier^index.
+	pow := NewFloat().SetInt(big.NewInt(int64(index)))
+	multiplier := BigFloatPow(feeBucketMultiplier, pow)
+	fee := NewFloat().Mul(feeBucketBaseRate, multiplier)
+
+	feeUint64, _ := fee.Uint64()
+	return feeUint64
+}
+
+// ComputeTimeBucketRangeFromIndex takes a fee range index and the GlobalParamsEntry, and returns the
+// [feeMin, feeMax] of the fee range.
+func ComputeTimeBucketRangeFromIndex(index int, globalParams *GlobalParamsEntry) (_min uint64, _max uint64) {
+	feeMin := computeTimeBucketMinFromIndex(index, globalParams)
+	feeMax := computeTimeBucketMinFromIndex(index+1, globalParams) - 1
+	return feeMin, feeMax
+}
+
+// ComputeTimeBucketIndexFromFeePerKBNanos takes a fee rate and the GlobalParamsEntry, and returns the fee range index.
+func ComputeTimeBucketIndexFromFeePerKBNanos(feePerKBNanos uint64, globalParams *GlobalParamsEntry) int {
+	feeBucketBaseRate, feeBucketMultiplier := computeFeeBucketBaseAndMultiplierFromGlobalParams(globalParams)
+
+	// Compute index = (log(feePerKBNanos) - log(feeBucketBaseRate)) / log(feeBucketMultiplier).
+	feeFloat := NewFloat().SetUint64(feePerKBNanos)
+	logFeeFloat := BigFloatLog(feeFloat)
```

Comment:
You can perform this check earlier before taking the logs of the base rate and multiplier
```suggestion
	// If the fee is less than the base rate, return 0.
	if feeFloat.Cmp(feeBucketBaseRate) < 0 {
		return 0
	}	
	logFeeFloat := BigFloatLog(feeFloat)
```